### PR TITLE
[NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44163 [NCCL] Add Environment Variable to guard Async Error Handling feature
* #41054 [NCCL] Destructor Blocks on WorkNCCL Completion
* #41053 [NCCL] WorkNCCL Helper Functions
* **#41052 [NCCL] Abort Errored and Timed Out NCCL Communicators from Watchdog Thread**
* #41051 [NCCL] Use cudaEventQuery to Poll for GPU operation errors
* #41050 [NCCL] Timeout Loop Thread for Async Error Handling

**This Commit:**
Watchdog Thread checks for error-ed or timed out WorkNCCL objects and aborts all associated NCCL Communicators. For now, we  also process these aborted communicators as with the existing Watchdog logic (by adding them to abortedCommIds and writing aborted communicator ids to the store.)

**This Stack:**
The purpose of this stack is to fix the hanging behavior observed in when using PyTorch DDP training with NCCL. In various situations (desynchronization, high GPU utilization, etc.), NCCL collectives may hang due to waiting on an unresponsive worker. This stack detects such hanging behavior and aborts timed-out collectives by throwing a user-visible exception, all with minimal perf regression. Training can then be restarted from a previous checkpoint with something like torchelastic.

Differential Revision: [D21943151](https://our.internmc.facebook.com/intern/diff/D21943151/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -478,29 +484,51 @@ void ProcessGroupNCCL::ncclCommWatchdogInternal() {
...
-                  buildNcclUniqueIdStr(ncclComm->getNcclId()));
-            }
+          LOG(INFO) << "Aborting communicators that received errors";
+          // We should not abort the communicators if we are performing a
+          // non-blocking wait(). The reason for this is that if we abort the
+          // nccl communicator, wait() might not throw exceptions and
+          // subsequent operations might run on garbage results.
+          // The current model is that when we call wait(), subsequent
+          // operations only run after this work is done or we hang forever
+          // waiting for the operation to complete.
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:494]: It seems like we violate the contract mentioned here if we remove `blockingWait_` here?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -478,29 +484,51 @@ void ProcessGroupNCCL::ncclCommWatchdogInternal() {
...
+                currentTimepoint - work->workStartTime_) > work->opTimeout_) {
+          std::exception_ptr exception_ptr = std::make_exception_ptr(
+              std::runtime_error("NCCL Operation Timed Out"));
+          work->setException(exception_ptr);
+          for (const auto& ncclComm : work->ncclComms_) {
+            ncclComm->ncclCommAbort();
+            abortedCommIds.emplace(buildNcclUniqueIdStr(ncclComm->getNcclId()));
           }
         }
       }
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:534]: This function is becoming pretty large with a multiple complex blocks, can we move each block out into separate helper functions for more clarity?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -478,29 +484,51 @@ void ProcessGroupNCCL::ncclCommWatchdogInternal() {
...
+        // communicators accordingly.
+        auto currentTimepoint = std::chrono::steady_clock::now();
+        if (std::chrono::duration_cast<std::chrono::milliseconds>(
+                currentTimepoint - work->workStartTime_) > work->opTimeout_) {
+          std::exception_ptr exception_ptr = std::make_exception_ptr(
+              std::runtime_error("NCCL Operation Timed Out"));
+          work->setException(exception_ptr);
+          for (const auto& ncclComm : work->ncclComms_) {
+            ncclComm->ncclCommAbort();
+            abortedCommIds.emplace(buildNcclUniqueIdStr(ncclComm->getNcclId()));
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:531]: Can we have some unit tests that actually verify this block of code does what it is supposed to?

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:542]: We have a test that verifies that this works in desync scenarios - we're working through the best way of including this as a unit test. We've also done larger training runs on Flow/TC that have desynchronizations to verify this behavior.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (jiayisuse) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:505]: I guess users will get the same behavior when `blockingWait_` is true? Verified on HPC trainer, blocking_wait just works with lower QPS

Comment (pritamdamania87) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: If `NCCL_BLOCKING_WAIT` is false and `NCCL_ASYNC_ERROR_HANDLING` is false, we would still end up aborting communicators here that might cause consistency issues where other ops after the aborted collective might run on corrupted data.

Comment (osalpekar) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: Good catch - thanks. We should probably guard this code block with `if (blockingWait_ || asyncErrorHandling_)` to handle this case.

**Pull request closed without merging**